### ISSUE TITLE
BinaryHeap NullPointerException

### ISSUE THREAD

**Issue by taluks** [libgdx/libgdx#5986]
#### Issue details
In prev release 7 march (gdx-1.9.11-SNAPSHOT)  my game worked fine

#### Reproduction steps/code
```java
BinaryHeap<Node> bh = new BinaryHeap<Node>();
bh.add(new Node(5f));
Node node = new Node(10f);
bh.add(node);
bh.remove(node);
bh.remove(node);
```
#### Version of LibGDX and/or relevant dependencies
gdx-1.9.11-SNAPSHOT
gdx-ai-1.8.2

#### Stacktrace
```java
java.lang.NullPointerException: 
  at com.badlogic.gdx.utils.BinaryHeap.down (BinaryHeap.java:156)
  at com.badlogic.gdx.utils.BinaryHeap.remove (BinaryHeap.java:102)
  at com.badlogic.gdx.ai.pfa.indexed.IndexedAStarPathFinder.visitChildren (IndexedAStarPathFinder.java:219)
  at com.badlogic.gdx.ai.pfa.indexed.IndexedAStarPathFinder.search (IndexedAStarPathFinder.java:117)
  at com.badlogic.gdx.ai.pfa.indexed.IndexedAStarPathFinder.searchNodePath (IndexedAStarPathFinder.java:94)
  at com.tpwar.game.maps.pathfinding.TiledPathFinder.find (TiledPathFinder.java:29)
  at com.tpwar.game.screens.MapScreen$InputAdapterExtension.touchUp (MapScreen.java:741)
  at com.badlogic.gdx.InputMultiplexer.touchUp (InputMultiplexer.java:124)
  at com.badlogic.gdx.backends.android.AndroidInput.processEvents (AndroidInput.java:406)
  at com.badlogic.gdx.backends.android.AndroidGraphics.onDrawFrame (AndroidGraphics.java:476)
  at android.opengl.GLSurfaceView$GLThread.guardedRun (GLSurfaceView.java:1581)
  at android.opengl.GLSurfaceView$GLThread.run (GLSurfaceView.java:1280)
```
![image](https://user-images.githubusercontent.com/20488327/78223250-3e454200-74cf-11ea-8620-d10ccd3b1c32.png)

#### Please select the affected platforms
- [X] Android
- [ ] iOS (robovm)
- [ ] iOS (MOE)
- [ ] HTML/GWT
- [X] Windows
- [ ] Linux
- [ ] MacOS

**Comment by EtK2000** [comment]
It would be helpful if you could supply a sample code that can be used to reproduce the issue.

**Comment by tommyettinger** [comment]
Are you using the current snapshot now? BinaryHeap was edited recently to fix an ordering bug.

**Comment by taluks** [comment]
https://github.com/taluks/libgdx-test
this test with 1.9.10 working

**Comment by tommyettinger** [comment]
I'm going to take a look; I'm curious what the change after March 7 could have done to trigger an NPE...

**Comment by tommyettinger** [comment]
This is pretty hard to follow. I think part of the issue has to do with the start and end nodes being non-walkable, maybe. It's hard to tell what coordinate offsets are in use because there's cases where identical names are used in different classes with no shared meaning, like how there's two `moveMap` arrays with different sizes. I printed this out of the JSON/TPMAP file; the origin is in the upper left, +1 x moves to the right, and +1 y moves down.

```
00000000000000000000
00000000000001000000
00000000000000100000
00000100000000000000
00000000000000001000
00001100000000110100
00001100011100110000
00010000111000000000
00000000011000000000
00000000000000000000
00001100000100110000
00011100000000110000
00000001000000000100
00000010000000000000
00000000000000000000
```

I'm pretty sure the `getNode(int, int)` method is indexed starting at 1, because the nodes it gets show different positions than the x and y given to getNode(). If that's correct, then the start and end nodes are both `1`, non-walkable. But I'm really not confident enough to say what the start and end actually are, here.

Still, a non-walkable end tile doesn't explain a Node being null inside BinaryHeap, especially if it worked before. I think I'll tinker with this a bit more, and see which positions cause an NPE.

**Comment by tommyettinger** [comment]
Yes! Not only can I confirm the bug, I can confirm a fix for the bug, and it only needs moving one line in BinaryHeap! The solution is to change, in BinaryHeap.remove(T):

```java
			nodes[node.index] = moved;
			nodes[size] = null;
```

to

```java
			nodes[size] = null;
			nodes[node.index] = moved;
```

I believe the issue happened when `size` and `node.index` were equal, so `nodes[node.index]` should have been `moved` but instead was `null`.

**Comment by NathanSweet** [comment]
Ah, that was tricky! Great find @tommyettinger!
